daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/IdeLedgerClient.scala

```diff
@@ -587,104 +583,74 @@ class IdeLedgerClient(
-            id,
-            throw new IllegalArgumentException(s"Unknown root node id $id"),
-          )
-          node match {
-            case create: Node.Create => ScriptLedgerClient.CreateResult(create.coid)
-            case exercise: Node.Exercise =>
-              ScriptLedgerClient.ExerciseResult(
-                exercise.templateId,
-                exercise.interfaceId,
-                exercise.choiceId,
-                exercise.exerciseResult.get,
-              )
-            case _: Node.Fetch | _: Node.LookupByKey | _: Node.Rollback =>
-              throw new IllegalArgumentException(s"Invalid root node: $node")
-          }
-        }
-        Right(transaction.roots.toSeq.map(convRootEvent))
-      case Left(ScenarioRunner.SubmissionError(err, tx)) =>
-        _currentSubmission = Some(ScenarioRunner.CurrentSubmission(optLocation, tx))
-        throw err
-    }
-
-  override def submitMustFail(
-      actAs: OneAnd[Set, Ref.Party],
-      readAs: Set[Ref.Party],
-      disclosures: List[Disclosure],
-      commands: List[command.ApiCommand],
-      optLocation: Option[Location],
-  )(implicit ec: ExecutionContext, mat: Materializer): Future[Either[Unit, Unit]] =
-    unsafeSubmit(actAs, readAs, disclosures, commands, optLocation).map {
-      case Right(ScenarioRunner.Commit(_, _, tx)) =>
-        _currentSubmission = Some(ScenarioRunner.CurrentSubmission(optLocation, tx))
-        Left(())
-      case Left(_) =>
-        _currentSubmission = None
-        _ledger = ledger.insertAssertMustFail(actAs.toSet, readAs, optLocation)
-        Right(())
-    }
-
-  override def submitTree(
-      actAs: OneAnd[Set, Ref.Party],
-      readAs: Set[Ref.Party],
-      disclosures: List[Disclosure],
-      commands: List[command.ApiCommand],
+      commands: List[ScriptLedgerClient.CommandWithMeta],
       optLocation: Option[Location],
+      languageVersionLookup: PackageId => Either[String, LanguageVersion],
+      errorBehaviour: ScriptLedgerClient.SubmissionErrorBehaviour,
   )(implicit
       ec: ExecutionContext,
       mat: Materializer,
-  ): Future[ScriptLedgerClient.TransactionTree] =
-    unsafeSubmit(actAs, readAs, disclosures, commands, optLocation).map {
-      case Right(ScenarioRunner.Commit(result, _, _)) =>
-        _currentSubmission = None
-        _ledger = result.newLedger
-        val transaction = result.richTransaction.transaction
-        def convEvent(id: NodeId): Option[ScriptLedgerClient.TreeEvent] =
-          transaction.nodes(id) match {
-            case create: Node.Create =>
-              Some(
-                ScriptLedgerClient.Created(
-                  create.templateId,
-                  create.coid,
-                  create.arg,
-                  blob(create, result.richTransaction.effectiveAt),
+  ): Future[Either[
+    ScriptLedgerClient.SubmitFailure,
+    (Seq[ScriptLedgerClient.CommandResult], Option[ScriptLedgerClient.TransactionTree]),
+  ]] = Future {
+    synchronized {
```

Comment:
IdeLedgerClient doesn't support concurrency.
Some notes here - the caller uses `Future.traverse`, which may call multiple submits on the same thread. (synchronized uses a ReentrantLock, so the same thread can enter multiple times.) As such, the full body of this submit is one Future computation, so the Future job scheduler cannot interrupt it, and violate our lock.